core/src/main/java/hudson/SystemProperties.java

```diff
@@ -0,0 +1,220 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package hudson;
+
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.servlet.ServletContext;
+
+/**
+ * Centralizes calls to System.getProperty() and related calls.
+ * This allows us to get values not just from environment variables but also from
+ * the ServletContext, so things like hudson.DNSMultiCast.disabled can be set
+ * in context.xml and the app server's boot script does not have to be changed.
+ * 
+ * While it looks like it on first glamce, this cannot be mapped to EnvVars.java
+ * because EnvVars.java is only for build variables, not Jenkins itself variables. 
+ * 
+ * This should be invoked for hudson parameters (e.g. hudson.DNSMultiCast.disabled),
+ * but not for system parameters (e.g. os.name).
+ * 
+ * @author Johannes Ernst
+ */
+public class SystemProperties {
+    /**
+     * The ServletContext to get the init parameters from.
+     */
+    private static ServletContext theContext;
+
+    /**
+     * Logger.
+     */
+    private static final Logger LOGGER = Logger.getLogger(SystemProperties.class.getName());
+
+    /**
+     * Gets the system property indicated by the specified key.
+     * This behaves just like System.getProperty(String), except that it
+     * also consults the ServletContext's init parameters.
+     * 
+     * @param      key   the name of the system property.
+     * @return     the string value of the system property,
+     *             or <code>null</code> if there is no property with that key.
+     *
+     * @exception  SecurityException  if a security manager exists and its
+     *             <code>checkPropertyAccess</code> method doesn't allow
+     *              access to the specified system property.
+     * @exception  NullPointerException if <code>key</code> is
+     *             <code>null</code>.
+     * @exception  IllegalArgumentException if <code>key</code> is empty.
+     */
+    public static String getProperty(String key) {
+        String value = System.getProperty(key); // keep passing on any exceptions
+        if( value != null ) {
+            if( LOGGER.isLoggable(Level.INFO )) {
+                LOGGER.log(Level.INFO, "Property (system): {0} => {1}", new Object[]{ key, value });
+            }
+        } else if( theContext != null ) {
+            value = theContext.getInitParameter(key);
+            if( value != null ) {
```

Comment:
I know we have some issues with inconsistent source code formatting, but… no. Just, no.